5 个 Kotlin 扩展技巧让你的 Android 代码更具表现力 (#8146)
xitu/gold-miner: article/2021/5-kotlin-extensions-to-make-your-android-code-more-expressive.md

> - 原文地址：[5 Kotlin Extensions To Make Your Android Code More Expressive](https://betterprogramming.pub/5-kotlin-extensions-to-make-your-android-code-more-expressive-4c9243cb9466)
> - 原文作者：[Siva Ganesh Kantamani](https://medium.com/@sgkantamani)
> - 译文出自：[掘金翻译计划](https://github.com/xitu/gold-miner)
> - 本文永久链接：[https://github.com/xitu/gold-miner/blob/master/article/2021/5-kotlin-extensions-to-make-your-android-code-more-expressive.md](https://github.com/xitu/gold-miner/blob/master/article/2021/5-kotlin-extensions-to-make-your-android-code-more-expressive.md)
> - 译者：[keepmovingljzy](https://github.com/keepmovingljzy)
> - 校对者：

# 5 个 Kotlin 扩展技巧让你的 Android 代码更具表现力

![](https://cdn-images-1.medium.com/max/12000/0*7wBRYeSryL8YT5u_)

你可能已经看过一堆关于 Kotlin 扩展相关的文章了，但这篇文章不仅仅只是关于扩展。它是关于如何让你的代码更具表现力，所以我专门讲解并概括我的顶级扩展，使代码尽可能自然一些。

本文的主要目的是学习如何使用扩展，而不是复制代码片段，以自己的方式表达代码。

## 介绍

Kotlin 是一种现代的、富有表现力的语言，它是为开发人员而构建的。在我看来，好的 Kotlin 代码就是能够以自然、可读的方式表达自己的代码。

> 从 Java 转换到 Kotlin 对我来说在很多方面都是一个巨大的转变，但我认为这在每个方面都是最好的。你可以参考 [我之前的文章](https://medium.com/better-programming/my-journey-from-java-to-kotlin-3bfcbcc6b734)。

Reviewer: ```suggestion
从使用 Java 转向使用 Kotlin 对我来说在很多方面都是一个巨大的转变，但我认为这在每个方面都是最好的。你可以参考 [我之前的文章](https://medium.com/better-programming/my-journey-from-java-to-kotlin-3bfcbcc6b734)。
```

Author: 你可以参考阅读一下，你不觉得这句很生硬吗

> 我最喜欢 Kotlin 的一点是扩展。作为一名移动 Java 开发人员，我从未想过向任何类添加自定义功能，尤其是向第三方库中的类添加自定义功能。但是当我听到扩展的概念时，它让我大吃一惊。对于 Android 开发者来说，这个特性打开了大量代码增强的大门。

Reviewer: ```suggestion
我最喜欢 Kotlin 的一点是扩展。作为一名移动端 Java 开发人员，我从未想过向任何类添加自定义功能，尤其是向第三方库中的类添加自定义功能。但是当我听到扩展的概念时，着实让我大吃一惊。对于 Android 开发者来说，这个特性打开了大量的通向代码增强的大门。
```

> “Kotlin 提供了新功能扩展类的能力，而无需继承该类或使用类似装饰器之类的设计模式。这是通过称为扩展的特殊声明来完成。” — [Kotlin 文档](https://kotlinlang.org/docs/extensions.html)

> 如果你想学习更多 Kotlin 扩展相关知识，阅读这篇文章： [使用 Kotlin 进行高级 Android 编程](https://medium.com/better-programming/advanced-android-programming-with-kotlin-5e40b1be22bb)

Reviewer: ```suggestion
如果你想学习更多 Kotlin 扩展相关知识，阅读这篇文章：[使用 Kotlin 进行高级 Android 编程](https://medium.com/better-programming/advanced-android-programming-with-kotlin-5e40b1be22bb)
```

我使用 Kotlin 扩展来使代码更具表现力，并使语言尽可能自然。

不要再拖了，开干！

> ## 1. 显示，隐藏，移除

Reviewer: ```suggestion
## 1. 显示，隐藏，移除（视图）
```

或许这里加个视图？更好？原文有些迷糊

> 移动开发人员的常见任务之一是隐藏和显示视图。 如果你使用 Java，你需要调用 `setVisibility` 方法传入`View.VISIBILE`或 `View.GONE` 来实现。 如下：

Reviewer: ```suggestion
移动开发人员的常见任务之一是隐藏和显示视图。如果你使用 Java，你需要调用 `setVisibility` 方法传入`View.VISIBILE`或 `View.GONE` 来实现。 如下：
```

```
view.setVisibility(View.GONE)
```

> 代码可以工作而且没有问题。但是使用 `set` 和 `get` 方法让它看起来更笨拙而不自然，Kotlin 提供了一种便利的方式来赋值，而不需要使用 `set` 和 `get` 方法。现有代码如下：

Reviewer: ```suggestion
上述的代码可以正常地工作。但是使用 `set` 和 `get` 方法让它看起来更笨拙而不自然，Kotlin 提供了一种更方便的方式来赋值，而不需要使用 `set` 和 `get` 方法。现有代码如下：
```

}
```

现在你可以使用如下方式：

```
view.show()
view.hide() 
view.remove()
```

现在看起来更友好更自然。我很乐意优化它，所以如果你有任何建议，请留下评论。

## 2. 校验

> 在任何开发环境中，验证字符串都是至关重要的。回到 2015 年，当我刚开始我的职业生涯，我看到一些应用程序显示`null`在一些文本字段中。这是因为没有适当的验证。

Reviewer: ```suggestion
在任何开发环境中，验证字符串都是至关重要的。回到 2015 年，当我刚开始我的职业生涯，我看到一些应用程序在一些文本字段中显示 `null`。这是因为没有适当的验证。
```

if(Utility.isStringNotNullOrEmpty(data))
```

> 从本质上讲数据类型不过是类。 因此我们可以使用 Kotlin 扩展来添加验证功能。 例如，我创建了以下 Kotlin 扩展，其数据类型为`String`，以检查其是否有效：

Reviewer: ```suggestion
从本质上讲数据类型不过是类。 因此我们可以使用 Kotlin 扩展来添加验证功能。 例如，我创建了以下 Kotlin 扩展，其数据类型为 `String`，以检查其是否有效：
```

if(data.valid())
```

> 显然， `data.valid()` 看起来比 Utility.isStringNotNullOrEmpty(data) 更简洁，可读性更好。调用数据类型上的扩展函数似乎比触发某些工具类函数更自然。下面是几个扩展，可以启发你编写自己的验证扩展：

Reviewer: ```suggestion
显然，`data.valid()` 看起来比 Utility.isStringNotNullOrEmpty(data) 更简洁，可读性更好。调用数据类型上的扩展函数似乎比触发某些工具类函数更自然。下面是几个扩展，可以启发你编写自己的验证扩展：
```

}
```

要了解内联函数和具体类型等高级特性，请参考[系列文章](https://medium.com/better-programming/advanced-android-programming-with-kotlin-5e40b1be22bb)。

现在让我们看看如何使用上面的扩展：

```Kotlin
val firstName by getValue<String>("firstName") // String?
val lastName by getValueNonNull<String>("lastName") // String
```

这种方式有三个优点：

1. 简洁，可读性好，代码量少。
2. 空安全。
3. 懒加载。

## 4. 资源扩展

> 在 Android 中，我们需要通过资源类访问项目资源。 这涉及到一些每次需要从资源文件中检索数据时都需要手动编写的样板代码。 如果没有任何扩展，检索 color 或 drawable 代码如下：

Reviewer: ```suggestion
在 Android 中，我们需要通过资源类访问项目资源。这涉及到一些每次需要从资源文件中检索数据时都需要手动编写的样板代码。如果没有任何扩展，检索 color 或 drawable 代码如下：
```

val drawable = R.drawable.launcher.asDrawable()
```

## 5. 显示 Alert Dialog, Toast, or Snackbar

> 当涉及到前端开发时，无论使用哪种平台，有时都需要向用户显示弹出框。 可能是用于显示不重要的数据，或者弹出提示用户确定或者显示一些错误。

Reviewer: ```suggestion
当涉及到前端开发时，无论使用哪种平台，有时都需要向用户显示弹出框。可能是用于显示不重要的数据，或者弹出提示用户确定或者显示一些错误。
```

}
```

> 编写这些扩展是一次性的工作。看看这些扩展的用法:

Reviewer: ```suggestion
编写这些扩展是一次性的工作。看看这些扩展的用法：
```
